### PR TITLE
feat: batch-invariant deterministic inference for DeepSeek MoE on FA4 + DeepGEMM + DeepEP

### PR DESCRIPTION
**Pull request by zyzshishui**
## Motivation

To make RL training reproducible, we want inference to be deterministic, and also extend&p/d invatiant for multi-turn. Currently, `--enable-deterministic-inference` does not support FA4 backend for B/GB GPUs.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py [hunk 1]
```diff
@@ -726,6 +726,33 @@ def forward_absorb_core(
                         topk_indices=topk_indices,
                     )
                     attn_output = fusion_plan.attn_output_buf
+                elif (
+                    self.current_attention_backend == "fa4"
+                    and forward_batch.forward_mode.is_decode_or_idle()
+                ):
+                    # TODO(Yuzhen): FA4's underlying kvcache kernel does not yet support in-place KV updates,
+                    # so write the MLA KV cache explicitly before invoking FA4 in read-only mode.
+                    get_token_to_kv_pool().set_mla_kv_buffer(
```
Comment (ch-wan) [inline python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py]: check https://github.com/zyzshishui/sglang/blob/main/python/sglang/srt/layers/attention/flashattention_backend.py#L959

File: python/sglang/srt/layers/sampler.py [hunk 2]
```diff
@@ -176,6 +180,16 @@ def forward(
                 # Standard path: do softmax and sample from probs.
                 logits.div_(sampling_info.temperatures)
 
+                if (
+                    return_logprob
+                    and self.enable_deterministic
+                    and logprobs_via_logsoftmax_kernel is None
+                    and not SGLANG_RETURN_ORIGINAL_LOGPROB
+                ):
+                    logprobs_via_logsoftmax_kernel = torch.nn.functional.log_softmax(
```
Comment (zyzshishui) [inline python/sglang/srt/layers/sampler.py:189]: @ch-wan  prefill logprobs are computed `via torch.nn.functional.log_softmax` here https://github.com/sgl-project/sglang/blob/40a522203c539f99b4656b871dc747990c580431/python/sglang/srt/layers/logits_processor.py#L710

The origional decode is using `torch.log` because the probs already exist, so it can save this op for non-deterministic path
